#34009 avro generic record to beam row conversion added support for a…

wollowizard opened the pull request:
#34009 avro generic record to beam row conversion added support for all logical types and conversions

The PR modifies the Avro extension, specifically enhancing how Avro logical types are handled when converting between Avro records and Beam rows. It introduces support for a broader range of Avro logical types (e.g., decimal, uuid, timestamp-micros, etc.), improves compatibility with Avro’s SpecificRecord and GenericRecord, and ensures proper serialization/deserialization of these types.
The primary goal is to make AvroUtils more robust in converting Avro records (both GenericRecord and SpecificRecord) to Beam rows by fully supporting Avro’s logical types. A new `convertLogicalType` method is introduced to handle logical type conversions dynamically using a GenericData instance. This method checks for registered conversions (e.g., DecimalConversion, UUIDConversion) and applies them to transform Avro data into Beam-compatible formats.

Instead of hardcoding conversions for specific logical types (e.g., TimestampMillis to ReadableInstant), the code now delegates to Avro’s GenericData conversion system. This makes it extensible to any logical type with a registered conversion, including custom ones.

Just fyi, my initial interest in this was given by the fact that I have some `SpecificRecord`s (subclass of `GenericRecord`) for which I have classes generated with the `avro-maven-plugin`. They have java time fields, uuids and bigdecimal fields and the generated code also adds conversion to the GenericData of the generated class. So it seems that the generated code is nice and complete but the conversion to a beam row failed because beam's AvroUtils' code does not support all logical types and has a predefined list of accepted input types (for example does not accept a GenericRecord with a BigDecimal field, it needs to be ByteBuffer, its raw type) . I thought a more flexible approach would be to use the avro conversions. 

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

wollowizard commented:
Run Java PreCommit

wollowizard commented:
assign set of reviewers

damccorm commented on sdks/java/extensions/avro/src/test/avro/org/apache/beam/sdk/extensions/avro/schemas/logicaltypes/logical-types.avsc:
```diff
@@ -0,0 +1,140 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
we don't need to include a license header here since we're not distributing this file (and it is under the RAT exclusion)

damccorm commented on buildSrc/build.gradle.kts:
```diff
@@ -45,7 +45,7 @@ dependencies {
 
   runtimeOnly("com.google.protobuf:protobuf-gradle-plugin:0.8.13")                                         // Enable proto code generation
-  runtimeOnly("com.github.davidmc24.gradle.plugin:gradle-avro-plugin:1.9.1")                               // Enable Avro code generation
+  runtimeOnly("com.github.davidmc24.gradle.plugin:gradle-avro-plugin:1.1.0")                               // Enable Avro code generation. Version 1.1.0 is the last supporting avro 1.10.2
```
I'm not sure about downgrading this dependency. Why do we specifically need to target 1.1.0? It is a rather old dependency, and downgrading could introduce other problems

wollowizard replied:
restored 1.9.1. My thinking is that since the earliest tested avro version is 1.10.2 (see `sdks/java/extensions/avro/build.gradle`) , it would be good to use a version of the avro plugin that uses avro 1.10.2.

In any case, the whole plugin seems no longer maintained, so maybe it would be a good idea to remove this altogether and just use the avro-tools jar, like elsewhere. seems like this could be addressed in a separate task though

damccorm commented on sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:
```diff
@@ -154,6 +158,15 @@ public class AvroUtils {
+    GENERIC_DATA_WITH_DEFAULT_CONVERSIONS = new GenericData();
+    addLogicalTypeConversions(GENERIC_DATA_WITH_DEFAULT_CONVERSIONS);
+    GENERIC_DATA_WITH_DEFAULT_CONVERSIONS.addLogicalTypeConversion(
+        new Conversions.DecimalConversion());
```
Is there a reason not to just add this to `addLogicalTypeConversion`

wollowizard replied:
in `addLogicalTypeConversions` I see this comment:
```
// do not add DecimalConversion by default as schema must have extra 'scale' and 'precision'
// properties. avro reflect already handles BigDecimal as string with the 'java-class' property
```
so I suppose that's the reason. I see that this method is only used in `org.apache.beam.sdk.extensions.avro.io.AvroDatumFactory.ReflectDatumFactory` so that seems relevant

damccorm commented:
Adding some folks who may have some more context here: @aromanenko-dev @Abacn

wollowizard commented:
@aromanenko-dev I think you might indeed be the right person to review this, as this is just an addition to the avro extension that as far as i can see was developed by you. I've added a description to the PR, but please do let me know if I can give additional explanation.

aromanenko-dev commented:
Sure, I'll take a look on this asap. 
Thanks!

aromanenko-dev commented:
@wollowizard Maybe this can be helpful (see subtasks):
https://github.com/apache/beam/issues/24292

Initial goal was to extract Avro from Beam's `core` and support different versions of Avro (even old ones) that have some incompatibility. So, most of the code was copy-pasted and adjusted for testing against different Avro versions:
https://github.com/apache/beam/pull/25216/files

Imho, if your changes are fine with this then it should be fine.

wollowizard commented:
> @wollowizard Maybe this can be helpful (see subtasks): #24292
> 
> Initial goal was to extract Avro from Beam's `core` and support different versions of Avro (even old ones) that have some incompatibility. So, most of the code was copy-pasted and adjusted for testing against different Avro versions: https://github.com/apache/beam/pull/25216/files
> 
> Imho, if your changes are aligned with this then it should be fine.

@aromanenko-dev that was indeed helpful, to get some context. I believe the change is aligned, and I have added extra tests to specifically check avro records for specific record classes created with 1.8.2 and 1.9.2

aromanenko-dev commented:
Run PostCommit_Java_Avro_Versions

aromanenko-dev commented:
Run Java Avro Versions PostCommit

aromanenko-dev commented:
@wollowizard There is a dedicated workflow to run tests against different Avro versions:
https://github.com/apache/beam/blob/master/.github/workflows/beam_PostCommit_Java_Avro_Versions.yml

wollowizard commented:
Run Java Avro Versions PostCommit

wollowizard commented:
@aromanenko-dev I am not sure how to proceed, would you mind giving another quick review? thanks in advance

aromanenko-dev reviewed (approved):
@wollowizard Well, I'm not on the project anymore for about one year, so, tbh I didn't follow any Avro extension changes. So, maybe ask another person who has more knowledge on this to make sure that it doesn't break the things.

Though, on the first sight, it LGTM in case if all checks are passing and, especially, multiple versions Avro checks. Then I think we are good.

wollowizard commented:
@damccorm @Abacn following the last comment, would you be able to take a look at this?

Abacn commented on sdks/java/extensions/avro/build.gradle:
```diff
@@ -67,6 +67,7 @@ dependencies {
   implementation library.java.error_prone_annotations
   implementation library.java.avro
   implementation library.java.joda_time
+  implementation 'org.apache.commons:commons-lang3'
```
please use `library.java.commons_lang3`

wollowizard commented:
@Abacn thanks, I've made the change and resolved your comment
